beacon-chain/rpc/beaconv1/state.go

```diff
@@ -56,3 +100,80 @@ func (bs *Server) GetStateFork(ctx context.Context, req *ethpb.StateRequest) (*e
 func (bs *Server) GetFinalityCheckpoints(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateFinalityCheckpointResponse, error) {
 	return nil, errors.New("unimplemented")
 }
+
+func (bs *Server) headStateRoot(ctx context.Context) ([]byte, error) {
+	stateRoot, err := bs.ChainInfoFetcher.HeadRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	}
+	return stateRoot, nil
+}
+
+func (bs *Server) genesisStateRoot(ctx context.Context) ([]byte, error) {
+	stateRoot, err := bs.ChainStartFetcher.PreGenesisState().HashTreeRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	}
+	return stateRoot[:], nil
+}
+
+func (bs *Server) finalizedStateRoot(ctx context.Context) ([]byte, error) {
+	var blockRoot [32]byte
+	copy(blockRoot[:], bs.ChainInfoFetcher.FinalizedCheckpt().Root)
+	state, err := bs.StateGenService.StateByRoot(ctx, blockRoot)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain state: %v", err)
+	}
+	stateRoot, err := state.HashTreeRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	}
+	return stateRoot[:], nil
+}
+
+func (bs *Server) justifiedStateRoot(ctx context.Context) ([]byte, error) {
+	var blockRoot [32]byte
+	copy(blockRoot[:], bs.ChainInfoFetcher.CurrentJustifiedCheckpt().Root)
+	state, err := bs.StateGenService.StateByRoot(ctx, blockRoot)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain state: %v", err)
+	}
+	stateRoot, err := state.HashTreeRoot(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to obtain root: %v", err)
+	}
+	return stateRoot[:], nil
+}
+
+func (bs *Server) stateRootByHex(ctx context.Context, stateId []byte) ([]byte, error) {
```

Comment:
Does this only work if I pass in a state root that the head state has collected? Not historically?

Reply:
I discussed this with Nishant and we decided to stick with the last 8192 states for now. Allowing to fetch any state would require a ton of work.

Reply:
To add onto this, its an issue with our current indexing of blocks and state in our db. We index all our states by block root, so as of the current moment no efficient method to retrieve it by state root from the db. We can look towards adding support for it in the future.( for now a limited feature set would be fine) unless there is demand from downstream consumers for far back states.